magpie/services.py

```diff
@@ -72,85 +65,315 @@ def __init__(self, service, request):
+        admin_group = get_constant("MAGPIE_ADMIN_GROUP", self.request)
+        admin_group = GroupService.by_group_name(admin_group, db_session=db_session)
+        if admin_group in user.groups:  # noqa
+            return [PermissionSet(perm, Access.ALLOW, Scope.MATCH, PermissionType.EFFECTIVE) for perm in permissions]
+
+        # current and parent resource(s) recursive-scope
+        match = allow_match
+        while resource is not None:  # bottom-up until service is reached
+
+            # include both permissions set in database as well as defined directly on resource
+            cur_res_perms = ResourceService.perms_for_user(resource, user, db_session=db_session)
+            cur_res_perms.extend(permission_to_pyramid_acls(resource.__acl__))
+
+            for perm_name in requested_perms:
+                for perm_tup in cur_res_perms:
+                    perm_set = PermissionSet(perm_tup)
+
+                    # if user is owner (directly or via groups), all permissions are set,
+                    # but continue processing this resource until end in case user explicit deny reverts it
+                    if perm_tup.perm_name == ALL_PERMISSIONS:
+                        for perm in requested_perms:
+                            all_perm = PermissionSet(perm, perm_set.access, perm.scope, perm.type)
+                            if perm_set.access == Access.DENY:
+                                effective_perms[perm] = all_perm
+                            else:
+                                effective_perms.setdefault(perm, all_perm)
+                    # skip if the current permission must not be processed (at all or for the moment until next iter)
+                    elif perm_set.name not in requested_perms or perm_set.name != perm_name:
+                        continue
+                    # only first resource can use match, parents are recursive-only
+                    if not match and perm_set.scope == Scope.MATCH:
+                        continue
+
+                    # user direct permissions have priority over inherited ones from groups
+                    if perm_set.type == PermissionType.DIRECT:
+                        perm = effective_perms.get(perm_name)
+                        # explicit user deny overrides user allow if any already found
+                        if perm is None or perm.type == PermissionType.INHERITED or perm_set.access == Access.DENY:
+                            effective_perms[perm_name] = perm_set
+                        continue  # final decision for this user, skip any group permissions
+
+                    # otherwise check if for another group permission and yet no user permission
+                    perm = effective_perms.get(perm_name)
+                    if perm is None or perm.type == PermissionType.INHERITED:
+                        # explicit deny overrides allow if any already found
+                        if perm is None or perm_set.access == Access.DENY:
```

Comment:
here, if I correctly interpret the ifs it should resolve to this:
if perm is None or (perm.type == PermissionType.INHERITED and perm_set.access == Access.DENY):
So here we set the permission if not already set : fine
or if the current permission is inherited and we got a deny : ok 
Since there is only 2 cases where the perm is set, do you think that the ifs could be simplified?

Reply:
Yes. It looks this way because of the above block that had a similar 2-tier ifs, but I don't mind. The interpretation is like you resolved it.